src/hook.c

```diff
@@ -0,0 +1,216 @@
+/*
+ * Copyright (C) the libgit2 contributors. All rights reserved.
+ *
+ * This file is part of libgit2, distributed under the GNU GPL v2 with
+ * a Linking Exception. For full terms see the included COPYING file.
+ */
+
+#include "common.h"
+#include "posix.h"
+#include "repository.h"
+
+#include "git2/hook.h"
+
+#define GIT_HOOKS_DIR "hooks/"
+
+const char * const githooks[] = {
+	"applypatch-msg",
+	"pre-applypatch",
+	"post-applypatch",
+	"pre-commit",
+	"prepare-commit-msg",
+	"commit-msg",
+	"post-commit",
+	"pre-rebase",
+	"post-checkout",
+	"post-merge",
+	"pre-push",
+	"pre-receive",
+	"update",
+	"post-receive",
+	"post-update",
+	"push-to-checkout",
+	"pre-auto-gc",
+	"post-rewrite",
+};
+
+int git_hook_dir(git_buf *out_dir, git_repository *repo)
+{
+	int err;
+	git_buf tmp_path = GIT_BUF_INIT;
+	git_buf cfg_path = GIT_BUF_INIT;
+	git_config *cfg;
+
+	assert(out_dir && repo);
+
+	/* We need to check for an override in the repo config */
+	err = git_repository_config__weakptr(&cfg, repo);
+	if (err != 0)
+		return err;
+
+	err = git_config_get_path(&cfg_path, cfg, "core.hooksPath");
+	if (err == GIT_ENOTFOUND) {
+		git_buf_puts(&tmp_path, repo->path_repository);
+		git_buf_joinpath(&tmp_path, tmp_path.ptr, GIT_HOOKS_DIR);
+	} else if (err == GIT_OK) {
+		git_buf_joinpath(&tmp_path, repo->path_repository, cfg_path.ptr);
+		git_path_resolve_relative(&tmp_path, 0);
+		git_buf_free(&cfg_path);
+	} else {
+		/* XXX: error reporting */
+		return NULL;
+	}
+
+	*out_dir = tmp_path;
```

Comment:
I'd rather make the folowing instead of doing a struct-copy here:

```C
git_buf_attach(out_dir, git_buf_cstr(&tmp_path), git_buf_len(&tmp_path));
```

You could also consider making the out-pointer point to a C string instead of to a buffer.